### PR TITLE
[v7r1] Allow to transfer from tape using different staging and transfer protocols (CTA)

### PR DESCRIPTION
**Pull request by chaen**
Up to now, when copying files from a tape system, you would always use `SRM` for everything: staging and copying. This is not true anymore with CTA that can only stage via xroot but transfer with more protocols.

This PR implements the split between the transfer/access protocols and the staging protocols, as well as the necessary machinery to perform the protocol translation with FTS. 
It should be totally transparent to existing configuration, and really only useful for CTA (today at least)
I won't comment much more, because more info are in the PR's doc

This is what a job looks like for example 

image

This PR is awaiting for 2 fixes on the FTS side as well as configuration changes on the CTA side to be completely tested.
Update 29.01: fixes have been implemented in FTS, however more surprises arise (for example, we could be limited to one file per fts job..)

**IMPORTANT** it will require a simple db update that I will add to the wiki
```
alter table Files CHANGE status status enum('New','Submitted','Ready','Active','Finished','Canceled','Staging','Failed','Defunct','Started', 'Not_used') DEFAULT 'New'; 
```

BEGINRELEASENOTES

*DMS
CHANGE: split staging and access protocols
NEW: FTS3 can submit transfers with staging protocols different from transfer protocols

ENDRELEASENOTES

### CODE REVIEW

**Comment by fstagni** [comment]
Just somewhat related, FYI: https://indico.cern.ch/event/876772/contributions/4175665/attachments/2170369/3664231/REST%20for%20Tape.pdf

**Comment by chaen** [comment]
Yep exactly. That's what the `(today at least)` comment in the PR description refers to :-) And this time, they seem to have realistic objectives (at least they had. Then some people started having ideas to make it another monster....) so I have good hope it will be in a reasonable timescale :-)

**Comment by chaen** [comment]
@andresailer you can start reviewing. I am still doing final tests, but I think it is pretty much final version

File: Resources/Storage/StorageElement.py [hunk 1]
```diff
@@ -690,7 +708,10 @@ def generateTransferURLsBetweenSEs(self, lfns, sourceSE, protocols=None):
         :param sourceSE: storageElement instance of the sourceSE
         :param protocols: ordered protocol restriction list
 
-        :return: dictionnary Successful/Failed with pair (src, dest) urls
+        :return: dictionnary with keys::
```
Comment (fstagni) [inline Resources/Storage/StorageElement.py:711]:
```suggestion
        :return: dictionary with keys::
```
:)

Comment (chaen) [reply Resources/Storage/StorageElement.py:711]: :-D The usual :-)

**Comment by fstagni** [comment]
@andresailer  the final review of this is for you!

File: DataManagementSystem/Client/FTS3Job.py [hunk 2]
```diff
@@ -277,29 +344,50 @@ def _constructTransferJob(self, pinTime, allLFNs, target_spacetoken, protocols=N
...
+
+      # Under normal circumstances, we simply submit an fts transfer as such:
+      # * srcProto://myFile -> destProto://myFile
+      #
+      # Even in case of the source storage being a tape system, it works fine.
+      # However, if the staging and transfer protocols are different (CTA),
+      #  we use the multihop machineryto submit two sequential fts transfers:
+      # one to stage, one to transfer.
+      # It looks like such
+      # * stageProto://myFile -> stageProto://myFile
```
Comment (atsareg) [inline DataManagementSystem/Client/FTS3Job.py:365]: Is it like that ?
```suggestion
      # * stageProto://myFile -> srcProto://myFile
```

File: DataManagementSystem/Client/FTS3Job.py [hunk 3]
```diff
@@ -266,6 +300,39 @@ def _constructTransferJob(self, pinTime, allLFNs, target_spacetoken, protocols=N
...
+    # This contains the staging URLs if they are different from the transfer URLs
+    # (CTA...)
+    allStageURLs = dict()
+
+    # In case we are transfering from a tape system, and the stage protocol
+    # is not the same as the transfer protocol, we generate the staging URLs
+    # to do a multihop transfer. See bellow.
+    if sourceIsTape:
+      srcProto, _destProto = res['Value']['Protocols']
+      if srcProto not in srcSE.localStageProtocolList:
```
Comment (andresailer) [inline DataManagementSystem/Client/FTS3Job.py:312]: I asked @chaen directly what srcProto and localStageProtocolList is for CTA, because I got lost

**Review by andresailer — CHANGES_REQUESTED** [review]
There are couple small things that need to be fixed.

File: Resources/Storage/CTAStorage.py [hunk 4]
```diff
@@ -21,6 +21,9 @@ class CTAStorage(GFAL2_XROOTStorage):
               BackendType = Cta
               AccessProtocols = root
               WriteProtocols = root
+              # This is very important if you want to transfer
+              # using a different protocol
+              StageProtocols = root
```
Comment (chaen) [inline Resources/Storage/CTAStorage.py:26]: to myself: add HTTPs config in the example

File: docs/source/AdministratorGuide/Resources/storage.rst [hunk 5]
```diff
@@ -311,6 +312,14 @@ Protocol matrix
 
 In order to make it easier to debug, the script :ref:`dirac-dms-protocol-matrix` will generate a CSV files that allows you to see what would happen if you were to try transfers between SEs
 
+
+Staging
+^^^^^^^
+
+Up to recently, any protocol that was defined as `AccessProtocol` was also used for staging (plot twister: only `srm` could do it).
```
[the remark this answers is not in the document]
Comment (chaen) [reply docs/source/AdministratorGuide/Resources/storage.rst:319]: Also double backtick

File: DataManagementSystem/Client/FTS3Job.py [hunk 6]
```diff
@@ -266,6 +300,39 @@ def _constructTransferJob(self, pinTime, allLFNs, target_spacetoken, protocols=N
 
     allSrcDstSURLs = res['Value']['Successful']
 
+    # This contains the staging URLs if they are different from the transfer URLs
+    # (CTA...)
+    allStageURLs = dict()
+
+    # In case we are transfering from a tape system, and the stage protocol
+    # is not the same as the transfer protocol, we generate the staging URLs
+    # to do a multihop transfer. See bellow.
```
[the remark this answers is not in the document]
Comment (chaen) [reply DataManagementSystem/Client/FTS3Job.py:309]: fixed

File: DataManagementSystem/Client/FTS3Job.py [hunk 7]
```diff
@@ -266,6 +289,28 @@ def _constructTransferJob(self, pinTime, allLFNs, target_spacetoken, protocols=N
...
+    if sourceIsTape:
+      srcProto, _destProto = res['Value']['Protocols']
+      if srcProto not in srcSE.localStageProtocolList:
+        res = srcSE.getURL(allSrcDstSURLs, protocol=srcSE.localStageProtocolList)
+
+        if not res['OK']:
+          return res
+
+        for lfn, reason in res['Value']['Failed'].iteritems():
```
[the remark this answers is not in the document]
Comment (chaen) [reply DataManagementSystem/Client/FTS3Job.py:329]: thanks, fixed

File: DataManagementSystem/Client/FTS3Job.py [hunk 8]
```diff
@@ -422,15 +511,17 @@ def _constructStagingJob(self, pinTime, allLFNs, target_spacetoken):
         continue
 
       sourceSURL = targetSURL = allTargetSURLs[ftsFile.lfn]
+      ftsFileID = getattr(ftsFile, 'fileID')
+      trans_metadata = {'desc': 'Stage %s' % ftsFileID, 'fileID': ftsFileID}
```
[the remark this answers is not in the document]
Comment (chaen) [reply DataManagementSystem/Client/FTS3Job.py:515]: You want it in the transfer or in the job, or in both ?

File: DataManagementSystem/Client/FTS3Job.py [hunk 9]
```diff
@@ -641,12 +732,15 @@ def _fillAccountingDict(self, jobStatusDict):
 
     accountingDict["TransferOK"] = len(successfulFiles)
     accountingDict["TransferTotal"] = len(filesInfoList)
-    accountingDict["TransferSize"] = sum([fileDict['filesize'] for fileDict in successfulFiles])
+    # We need this if in the list comprehension because staging only jobs have `None`
+    # as filesize
```
[the remark this answers is not in the document]
Comment (chaen) [reply DataManagementSystem/Client/FTS3Job.py:736]: yup

File: DataManagementSystem/Client/FTS3Job.py [hunk 10]
```diff
@@ -641,12 +732,15 @@ def _fillAccountingDict(self, jobStatusDict):
...
+    accountingDict["TransferSize"] = sum([fileDict['filesize'] for fileDict in successfulFiles if fileDict['filesize']])
     accountingDict["FinalStatus"] = self.status
     accountingDict["Source"] = sourceSE
     accountingDict["Destination"] = targetSE
+    # Same as above for the condition
```
[the remark this answers is not in the document]
Comment (chaen) [reply DataManagementSystem/Client/FTS3Job.py:741]: done

File: DataManagementSystem/Client/FTS3Operation.py [hunk 11]
```diff
@@ -428,6 +432,43 @@ def prepareNewJobs(self, maxFilesPerJob=100, maxAttemptsPerFile=10):
...
+    """ Checks whether transfers between the two SE given as parameters
+        need a multi hop transfer to stage with a different protocol
+        than the transfer one.
+
+        :param str sourceSEName: source storage element name
+        :param str destSEName: destination storage element name
+
+        :returns: boolean
+    """
+    srcSE = StorageElement(sourceSEName, vo=self._vo)
```
[the remark this answers is not in the document]
Comment (chaen) [reply DataManagementSystem/Client/FTS3Operation.py:445]: well spotted, fixing it

File: Resources/Storage/StorageElement.py [hunk 12]
```diff
@@ -690,7 +708,10 @@ def generateTransferURLsBetweenSEs(self, lfns, sourceSE, protocols=None):
         :param sourceSE: storageElement instance of the sourceSE
         :param protocols: ordered protocol restriction list
 
-        :return: dictionnary Successful/Failed with pair (src, dest) urls
+        :return: dictionary with keys::
```
[the remark this answers is not in the document]
Comment (chaen) [reply Resources/Storage/StorageElement.py:711]: ok, done

File: DataManagementSystem/Client/FTS3Job.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (chaen) [reply DataManagementSystem/Client/FTS3Job.py:517]: Added in the Job

**Review by andresailer — COMMENTED** [review]
Just some proposals for changes in the comments. What had me initially confused, was that I thought CTA always has stageprotocol != transferprotocol. That this is not the case is explained in the storage.rst, but some of the existing comments might make one think that, so I propose the modifications

**Review by andresailer — APPROVED** [review]

**Comment by chaen** [comment]
@andresailer  Thanks for the improvements, I committed them, please reaprove

File: Resources/Storage/CTAStorage.py [hunk 13]
```diff
@@ -21,17 +21,30 @@ class CTAStorage(GFAL2_XROOTStorage):
               BackendType = Cta
               AccessProtocols = root
               WriteProtocols = root
+              # This is very important if you have to stage with this protocol,  but might transfer
```
Comment (andresailer) [inline Resources/Storage/CTAStorage.py:24]:
```suggestion
              # This is very important if you have to stage with this protocol,  but might transfer
```

**Review by andresailer — APPROVED** [review]

**Comment by chaen** [comment]
@atsareg it would be great if you could quickly do a release with that one in, using DIRACOS 1.17 (the latest). I'd like to deploy it early next week. Thanks !

**Pull request merged**